File: packages/nodx-py/src/nodx/ast.py
def empty_attrs():
    return {"attrs": {}, "classes": [], "id": None}


def node(type_, attrs, children, inlines, text):
    return {
        "attrs": attrs["attrs"],
        "children": children,
        "classes": attrs["classes"],
        "id": attrs["id"],
        "inlines": inlines,
        "text": text,
        "type": type_,
    }


def first_heading(nodes):
    for item in nodes:
        if item["type"] == "heading":
            return plain_node_text(item)
        child = first_heading(item["children"])
        if child:
            return child
    return None


def plain_node_text(item):
    parts = [plain_inlines(item["inlines"])]
    for child in item["children"]:
        text = plain_node_text(child)
        if text:
            parts.append(text)
    return " ".join(part for part in parts if part)


def plain_inlines(inlines):
    out = ""
    for item in inlines:
        type_ = item["type"]
        if type_ in ("text", "code"):
            out += item["text"]
        elif type_ == "math-inline":
            out += item["source"]
        elif type_ in ("strong", "em", "mark", "sub", "sup"):
            out += plain_inlines(item["children"])
        elif type_ == "link":
            out += plain_inlines(item["label"])
        elif type_ == "span":
            out += plain_inlines(item["children"])
        elif type_ == "var":
            out += "{{" + item["namespace"] + "." + item["name"] + "}}"
        elif type_ in ("ref", "footnote-ref", "citation-ref"):
            out += item["target"]
        elif type_ == "mention":
            out += "@" + item["kind"] + ":" + item["target"]
    return out

File: packages/nodx-js/src/ast.mjs
export function node(type, attrs, children, inlines, text) {
  const out = { attrs: attrs.attrs, children, classes: attrs.classes, id: attrs.id, inlines, text, type };
  if (Object.keys(attrs.styles ?? {}).length) out.styles = attrs.styles;
  return out;
}

export function emptyAttrs() {
  return { attrs: {}, classes: [], id: null, styles: {} };
}

export function firstHeading(nodes) {
  for (const item of nodes) {
    if (item.type === "heading") return plainNodeText(item);
    const child = firstHeading(item.children);
    if (child) return child;
  }
  return null;
}

export function plainNodeText(item) {
  const parts = [plainInlines(item.inlines)];
  for (const child of item.children) {
    const text = plainNodeText(child);
    if (text) parts.push(text);
  }
  return parts.filter(Boolean).join(" ");
}

export function plainInlines(inlines) {
  let out = "";
  for (const item of inlines) {
    switch (item.type) {
      case "text":
      case "code":
        out += item.text;
        break;
      case "math-inline":
        out += item.source;
        break;
      case "strong":
      case "em":
      case "mark":
      case "strike":
      case "sub":
      case "sup":
        out += plainInlines(item.children);
        break;
      case "link":
        out += plainInlines(item.label);
        break;
      case "span":
        out += plainInlines(item.children);
        break;
      case "var":
        out += "{{" + item.namespace + "." + item.name + "}}";
        break;
      case "ref":
      case "footnote-ref":
      case "citation-ref":
        out += item.target;
        break;
      case "mention":
        out += "@" + item.kind + ":" + item.target;
        break;
      case "line-break":
        out += " ";
        break;
    }
  }
  return out;
}
